<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<p>Syllabus for GEN220: High Throughput Biological Data Processing</p>
<h1 id="course-description">Course Description</h1>
<p>This course focuses on computational skills for processing data using programming language Python and UNIX environment. No prior programming experience is required, but some basic computer skills will be useful.</p>
<p>With the advancement of high throughput data generation methods, a major challenge that graduate students in life sciences have to face today is to analyze large amount of biological data. The objective of this course is to provide an opportunity for graduate students with no computer science background to learn the basic skills of handling high throughput biological data. It covers the Linux/Unix environment and the importance of the com1mand line interface; the Python programming language; program design, implementation, and testing; BioPython; Strategies for analyzing genome resequencing, RNASeq, sequencing data. Students build hands-on skills by analyzing real high throughput biological data through homework assignments and team projects.</p>
<p>Units: 3</p>
<p>Instructor: Jason Stajich (jason.stajich@ucr.edu)</p>
<p>Time and location: W 4:00-4:50PM, F 3:00-4:50PM, ULB104</p>
<p>Office Hours: By Appointment, 1207K Genomics</p>
<p><a href="https://biodataprog.github.io/GEN220/" class="uri">https://biodataprog.github.io/GEN220/</a></p>
<h2 id="prerequisites">Prerequisites</h2>
<ul>
<li>Coursework in genetics or molecular biology or permission of instructor</li>
</ul>
<h2 id="resources">Resources</h2>
<p>None of these texts are required for completion of the course but they will provide a great deal of helpful background and examples that will improve your ability to master UNIX or Programming in Python.</p>
<ol type="1">
<li><p><em>Bioinformatics Data Skills: Reproducible and Robust Research with Open Source Tools</em>. Vince Buffalo. 2015 O'Reilly &amp; Associates. Available from <a href="http://shop.oreilly.com/product/0636920030157.do">O'Reilly and Associates</a>, <a href="http://amazon.com/Bioinformatics-Data-Skills-Reproducible-Research/dp/1449367372">Amazon</a></p></li>
<li><p><em>Unix and Perl to the Rescue: A Primer</em>. Keith Bradnam and Ian Korf. <a href="http://korflab.ucdavis.edu/unix_and_Perl/">Unix and Perl Primer for Biologists</a></p></li>
<li><p><em>Unix and Perl to the rescue!</em> Bradnam and Korf. <a href="https://www.amazon.com/gp/product/0521169828?tag=keithbradnamc-20">Amazon</a></p></li>
<li><p><a href="http://rosalind.info/problems/locations/">Rosalind</a> - An online platform to learn bioinformatics and programming in Python.</p></li>
<li><p>Software Carpentry - <a href="https://software-carpentry.org/" class="uri">https://software-carpentry.org/</a> and Data Carpentry - <a href="http://www.datacarpentry.org/" class="uri">http://www.datacarpentry.org/</a>.</p></li>
<li><p>Berk Ekmekci, Charles E. McAnany, Cameron Mura. An Introduction to Programming for Bioscientists: A Python-Based Primer. PLoS Comp Bio. DOI: <a href="https://doi.org/10.1371/journal.pcbi.1004867">10.1371/journal.pcbi.1004867</a></p></li>
</ol>
<h2 id="grading">Grading</h2>
<ul>
<li>Programming Homework assignments (5 in total): 50% of grade</li>
<li>Team project: 50% of grade</li>
</ul>
<h2 id="homework">Homework</h2>
<ul>
<li><p>Homework is due before class on the date listed in the syllabus or if amended when assignment is given.</p></li>
<li><p>There will be a programming assignment every two weeks during the first half of the course. Programming assignments must be prepared along with any necessary input files or documentation to demonstrate program usage.</p></li>
<li><p>Code should be runnable as turned in. You will deposit your code in your github repository or if not possible, by iLearn. You can make one private personal repository to deposit and should organize a folder for each homework assignment (e.g. hw1, hw2, hw3).</p></li>
</ul>
<h2 id="projects">Projects</h2>
<ul>
<li><p>Project Topics will be discussed in October and teams will select a project idea to focus on.</p></li>
<li><p>Project will be 2-3 individuals working together.</p></li>
<li><p>A presentation will be made by each team - last day(s) of class.</p></li>
<li><p>A final report with the details will be turned in by the group.</p></li>
<li><p>The report needs to detail what each person's contribution is to the project.</p></li>
</ul>
<h2 id="schedule">Schedule</h2>
<table>
<tbody>
<tr class="odd">
<td>| Date | Day | Lecture Topic | Notes</td>
</tr>
<tr class="even">
<td>| :------ | :---- | :---------------------- | :------------ |</td>
</tr>
<tr class="odd">
<td>| Sep-27 | F | Course Intro / UNIX I: Cmdline, GitHub | |</td>
</tr>
<tr class="even">
<td>| Oct-2 | W | UNIX II: Biocluster HPCC, Running programs | |</td>
</tr>
<tr class="odd">
<td>| Oct-4 | F | UNIX III: Tools for data processing | HW1 Due |</td>
</tr>
<tr class="even">
<td>| Oct-9 | W | UNIX IV: Advanced UNIX and data processing | |</td>
</tr>
<tr class="odd">
<td>| Oct-11 | F | Python 1 - Variables, running, cmdline, strings, math | |</td>
</tr>
<tr class="even">
<td>| Oct-16 | W | Python 2 - Logic, loops, lists, iterator | |</td>
</tr>
<tr class="odd">
<td>| Oct-18 | F | Python 3 - I/O reading/writing files, directories | HW2 Due |</td>
</tr>
<tr class="even">
<td>| Oct-23 | W | Python 4 - Dictionaries, Arrays, functions | |</td>
</tr>
<tr class="odd">
<td>| Oct-25 | F | Python 5 - Structured data (CSV, XML, GFF, BED) | |</td>
</tr>
<tr class="even">
<td>| Oct-30 | W | Python 6 - Panda, BioPython | |</td>
</tr>
<tr class="odd">
<td>| Nov-1 | F | Alignment and Bioinformatics Algorithms | HW3 Due |</td>
</tr>
<tr class="even">
<td>| Nov-6 | W | Bioinformatics 1 - BLAST, cmdline &amp; automation | |</td>
</tr>
<tr class="odd">
<td>| Nov-8 | F | Bioinformatics 2 - Aligning short reads, coverage, identifying variants | |</td>
</tr>
<tr class="even">
<td>| Nov-13 | W | Bioinformatics 3 - Genome Assembly | |</td>
</tr>
<tr class="odd">
<td>| Nov-15 | F | Data Plotting and R graphics | HW4 Due |</td>
</tr>
<tr class="even">
<td>| Nov-20 | W | Bioinformatics 4 - RNASeq analyses | |</td>
</tr>
<tr class="odd">
<td>| Nov-22 | F | Bioinformatics 5 - Protein Sequence analyses | |</td>
</tr>
<tr class="even">
<td>| Nov-27 | W | Bioinformatics 6 - Microbiome | |</td>
</tr>
<tr class="odd">
<td>| <em>Nov-29</em> | F | <em>Thanksgiving Holiday</em> | |</td>
</tr>
<tr class="even">
<td>| Dec-4 | W | Bioinformatics 7 - Phylogenentic analyses | HW5 Due |</td>
</tr>
<tr class="odd">
<td>| Dec-6 | F | Presentations | |</td>
</tr>
<tr class="even">
<td>| Dec-? | ? | Extra Presentations | as needed |</td>
</tr>
</tbody>
</table>
</body>
</html>
